modules/ml/src/main/java/org/apache/ignite/ml/environment/LearningEnvironmentBuilder.java

```diff
@@ -17,109 +17,135 @@
+        return buildForWorker(-1);
     }
 
     /**
-     * Specifies Parallelism Strategy for LearningEnvironment.
+     * Specifies dependency (partition -> Parallelism Strategy Type for LearningEnvironment).
      *
-     * @param stgy Parallelism Strategy.
+     * @param stgyType Function describing dependency (partition -> Parallelism Strategy Type).
+     * @return This object.
      */
-    public <T> LearningEnvironmentBuilder withParallelismStrategy(ParallelismStrategy stgy) {
-        this.parallelismStgy = stgy;
+    public LearningEnvironmentBuilder withParallelismStrategyType(
+        IgniteFunction<Integer, ParallelismStrategy.Type> stgyType);
 
-        return this;
+    /**
+     * Specifies Parallelism Strategy Type for LearningEnvironment. Same strategy type will be used for all partitions.
+     *
+     * @param stgyType Parallelism Strategy Type.
+     * @return This object.
+     */
+    public default LearningEnvironmentBuilder withParallelismStrategyType(ParallelismStrategy.Type stgyType) {
+        return withParallelismStrategyType(constant(stgyType));
     }
 
     /**
-     * Specifies Parallelism Strategy for LearningEnvironment.
+     * Specifies dependency (partition -> Parallelism Strategy for LearningEnvironment).
      *
-     * @param stgyType Parallelism Strategy Type.
+     * @param stgy Function describing dependency (partition -> Parallelism Strategy).
+     * @return This object.
+     */
+    public LearningEnvironmentBuilder withParallelismStrategy(IgniteFunction<Integer, ParallelismStrategy> stgy);
+
+    /**
+     * Specifies Parallelism Strategy for LearningEnvironment. Same strategy type will be used for all partitions.
+     *
+     * @param stgy Parallelism Strategy.
+     * @return This object.
+     */
+    public default LearningEnvironmentBuilder withParallelismStrategy(ParallelismStrategy stgy) {
+        return withParallelismStrategy(constant(stgy));
+    }
+
+
+    /**
+     * Specify dependency (partition -> logging factory).
+     *
+     * @param loggingFactory Function describing (partition -> logging factory).
+     * @return This object.
+     */
+    public LearningEnvironmentBuilder withLoggingFactory(IgniteFunction<Integer, MLLogger.Factory> loggingFactory);
+
+    /**
+     * Specify logging factory.
+     *
+     * @param loggingFactory Logging factory.
+     * @return This object.
      */
-    public LearningEnvironmentBuilder withParallelismStrategy(ParallelismStrategy.Type stgyType) {
-        switch (stgyType) {
-            case NO_PARALLELISM:
-                this.parallelismStgy = NoParallelismStrategy.INSTANCE;
-                break;
-            case ON_DEFAULT_POOL:
-                this.parallelismStgy = new DefaultParallelismStrategy();
-                break;
-        }
-        return this;
+    public default LearningEnvironmentBuilder withLoggingFactory(MLLogger.Factory loggingFactory) {
+        return withLoggingFactory(constant(loggingFactory));
```

Comment:
Doesn't look like it's correct. If factory is not serializable you can't wrap it into function and so that make it serializable.